confgenerator/resourcedetector/gce_detector.go

```diff
@@ -0,0 +1,125 @@
+package resourcedetector
+
+import (
+	"fmt"
+	"strings"
+)
+
+// Keep synced with the Prometheus repo
+const (
+	metaLabelPrefix       = "__meta_"
+	gceLabel              = metaLabelPrefix + "gce_"
+	gceLabelProject       = gceLabel + "project"
+	gceLabelZone          = gceLabel + "zone"
+	gceLabelNetwork       = gceLabel + "network"
+	gceLabelSubnetwork    = gceLabel + "subnetwork"
+	gceLabelPublicIP      = gceLabel + "public_ip"
+	gceLabelPrivateIP     = gceLabel + "private_ip"
+	gceLabelInstanceID    = gceLabel + "instance_id"
+	gceLabelInstanceName  = gceLabel + "instance_name"
+	gceLabelTags          = gceLabel + "tags"
+	gceLabelMetadata      = gceLabel + "metadata_"
+	gceLabelLabel         = gceLabel + "label_"
+	gceLabelMachineType   = gceLabel + "machine_type"
+	gceLabelInterfaceIPv4 = gceLabel + "interface_ipv4_"
+)
+
+// The data provider interface for GCE environment
+// Implementation of this provider can use either the metadata server on VM,
+// or the cloud API
+type GCEDataProvider interface {
+	getProject() (string, error)
+	getZone() (string, error)
+	getNetwork() (string, error)
+	getSubnetwork() (string, error)
+	getPublicIP() (string, error)
+	getPrivateIP() (string, error)
+	getInstanceID() (string, error)
+	getInstanceName() (string, error)
+	getTags() (string, error)
+	getLabel() (string, error)
+	getMachineType() (string, error)
+	getMetadataKeys() (string, error)
+	getMetadata(key string) (string, string, error)
+	getInterfaceNames() (string, error)
+	getInterfaceIPv4(name string) (string, string, error)
+}
+
+// Nested property takes two steps: first to get the list of keys,
+// and then use the key to get individual values.
+// Note that the key are usually used to construct the final label
+type nestedProperties struct {
+	first func(GCEDataProvider) (string, error)
+	then  func(GCEDataProvider, string) (string, string, error)
+}
+
+// List of single value labels (non-nested)
```

Comment:
`labels` feels like a prometheus specific thing to me. It can get confusing especially since the metadata has its own `labels` too. Naming it instead to something like `metadata` or `attributes` might make it clearer.